testbed/testbed/results.go

```
@@ -45,52 +56,183 @@ type TestResult struct {
+		log.Fatalf(err.Error())
+	}
+	r.resultsFile, err = os.Create(path.Join(r.resultsDir, "CORRECTNESSRESULTS.md"))
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+
+	// Write the header
+	_, _ = io.WriteString(r.resultsFile,
+		"# Test Results\n"+
+			fmt.Sprintf("Started: %s\n\n", time.Now().Format(time.RFC1123Z))+
+			"Test                                    |Result|Duration|Sent Items|Received Items|Failure Count|Failures\n"+
+			"----------------------------------------|------|-------:|---------:|-------------:|------------:|--------\n")
+}
+
+func (r *CorrectnessResults) Add(testName string, result interface{}) {
+	testResult, ok := result.(*CorrectnessTestResult)
+	if !ok {
+		return
+	}
+	consolidated := consolidaateAssertionFailures(testResult.assertionFailures)
+	failuresStr := ""
+	for _, af := range consolidated {
+		failuresStr = fmt.Sprintf("%s%s,%#v!=%#v,count=%d; ", failuresStr, af.fieldPath, af.expectedValue,
+			af.actualValue, af.sumCount)
+	}
+	_, _ = io.WriteString(r.resultsFile,
+		fmt.Sprintf("%-40s|%-6s|%7.0fs|%10d|%14d|%13d|%s\n",
+			testResult.testName,
+			testResult.result,
+			testResult.duration.Seconds(),
+			testResult.sentSpanCount,
+			testResult.receivedSpanCount,
+			testResult.assertionFailureCount,
+			failuresStr,
 		),
 	)
-	r.totalDuration = r.totalDuration + result.duration
+	r.perTestResults = append(r.perTestResults, testResult)
+	r.totalAssertionFailures = r.totalAssertionFailures + testResult.assertionFailureCount
+	r.totalDuration = r.totalDuration + testResult.duration
+}
+
+func (r *CorrectnessResults) Save() {
+	_, _ = io.WriteString(r.resultsFile,
+		fmt.Sprintf("\nTotal assertion failures: %d\n", r.totalAssertionFailures))
+	_, _ = io.WriteString(r.resultsFile,
+		fmt.Sprintf("\nTotal duration: %.0fs\n", r.totalDuration.Seconds()))
+	r.resultsFile.Close()
+}
+
+func createDirectoryIfNeeded(resultsDir string) error {
+	dirPresent, err := exists(resultsDir)
+	if !dirPresent {
+		os.MkdirAll(resultsDir, os.ModePerm)
```

Returned error is not handled.

---

Fixed